### DISCUSSION TITLE
How to assert existence/ownership/permissions of directory while ignoring contents?

### DISCUSSION THREAD

**Discussion by catniptwinz** [CyberShadow/aconfmgr#185] [Q&A]
I'm experimenting with aconfmgr in the hopes of using it to manage some existing systems and I'd like to get some clarification regarding some behavior I'm observing from `aconfmgr save`.

I would like to assert that a directory should exist with specific ownership and permissions, but aconfmgr should not manage its contents. The eventual use case is a maildir used by system services.

Here is a minimal working example which reproduces the issue I'm encountering.

```
# mkdir -p /foo/bar/baz
# echo 'qux' > /foo/bar/baz/qux
# chmod -R 700 /foo/bar
# chown -R foo:foo /foo
```

`aconfmgr save` creates the assertions I'd expect to see:

```
CopyFile /foo/bar/baz/qux 700 foo foo
SetFileProperty /foo/bar/baz group foo
SetFileProperty /foo/bar/baz mode 700
SetFileProperty /foo/bar/baz owner foo
SetFileProperty /foo/bar group foo
SetFileProperty /foo/bar mode 700
SetFileProperty /foo/bar owner foo
SetFileProperty /foo group foo
SetFileProperty /foo owner foo
```

Now, we want to ignore the contents of `/foo/bar`. Add:

```
IgnorePath '/foo/bar/*'
CreateDir /foo/bar 700 foo foo
```

And `aconfmgr save` emits:

```
RemoveFile /foo/bar
RemoveFile /foo
SetFileProperty /foo/bar group ''
SetFileProperty /foo/bar mode ''
SetFileProperty /foo/bar owner ''
```

Why is this, and what is the correct way to assert the existence/ownership/permissions of a directory while ignoring its contents?

**Comment by CyberShadow** [comment]
I think `IgnorePath '/foo/bar/?*'` should work for ignoring a directory's contents but not the directory itself. Does that work?

**Reply by catniptwinz** [reply]
Unfortunately, no. The results are the same as with `IgnorePath '/foo/bar/*'`.

**Reply by CyberShadow** [reply]
Ah, this behavior is described here:

https://github.com/CyberShadow/aconfmgr/blob/e7f85b592375e50b25c324776a1e6d73d9c98490/src/common.bash#L380-L390

Needs some thought on how to proceed...

**Reply by catniptwinz** [reply]
Okay, I see. Indeed creating a non-ignored file inside the directory as a workaround seems to do what I need for the moment.

Intuitively I'd have hoped that the explicit `CreateDir` would have been sufficient to express "yes, we really want this directory even though it's empty", but I can see that the implementation wouldn't necessarily be straightforward.
